File: packages/netbridge/forwarder/protocol/frame.go
// Package protocol is the Go side of the NetBridge mux wire format. It mirrors the frame codec in the
// TypeScript @sw/netbridge package (the source of truth): one WebSocket message carries one frame — a
// 1-byte type, a 4-byte stream id, and a type-specific tail:
//
//	Open  [1][streamId:4][port:2][host…]  — a new TCP stream to host:port (name resolved at the exit)
//	Data  [2][streamId:4][bytes…]         — payload for an open stream
//	Close [3][streamId:4]                 — the stream is finished in the sender's direction
//
// The forwarder only encodes (it opens streams and streams data) and decodes Data/Close from the control
// plane; the channel envelope of @sw/netbridge is the control plane's concern and never reaches here.
package protocol

import (
	"encoding/binary"
	"errors"
)

type FrameType byte

const (
	FrameOpen  FrameType = 1
	FrameData  FrameType = 2
	FrameClose FrameType = 3
)

// MaxDataBytes caps a Data frame's payload so a single WebSocket message never grows unbounded; the
// forwarder chunks a TCP stream into pieces no larger than this. Matches @sw/netbridge.
const MaxDataBytes = 64 * 1024

const (
	headerSize   = 5
	openPortSize = 2
)

var ErrMalformedFrame = errors.New("netbridge: malformed frame")

type Frame struct {
	Type     FrameType
	StreamID uint32
	Host     string // Open only
	Port     uint16 // Open only
	Payload  []byte // Data only
}

func EncodeOpen(streamID uint32, host string, port uint16) []byte {
	buffer := make([]byte, headerSize+openPortSize+len(host))

	writeHeader(buffer, FrameOpen, streamID)
	binary.BigEndian.PutUint16(buffer[headerSize:], port)
	copy(buffer[headerSize+openPortSize:], host)

	return buffer
}

func EncodeData(streamID uint32, payload []byte) []byte {
	buffer := make([]byte, headerSize+len(payload))

	writeHeader(buffer, FrameData, streamID)
	copy(buffer[headerSize:], payload)

	return buffer
}

func EncodeClose(streamID uint32) []byte {
	buffer := make([]byte, headerSize)

	writeHeader(buffer, FrameClose, streamID)

	return buffer
}

func Decode(message []byte) (Frame, error) {
	if len(message) < headerSize {
		return Frame{}, ErrMalformedFrame
	}

	frame := Frame{Type: FrameType(message[0]), StreamID: binary.BigEndian.Uint32(message[1:])}

	switch frame.Type {
	case FrameOpen:
		return decodeOpen(message, frame)
	case FrameData:
		frame.Payload = message[headerSize:]

		return frame, nil
	case FrameClose:
		return frame, nil
	default:
		return Frame{}, ErrMalformedFrame
	}
}

func decodeOpen(message []byte, frame Frame) (Frame, error) {
	if len(message) < headerSize+openPortSize {
		return Frame{}, ErrMalformedFrame
	}

	frame.Port = binary.BigEndian.Uint16(message[headerSize:])
	frame.Host = string(message[headerSize+openPortSize:])

	if frame.Host == "" {
		return Frame{}, ErrMalformedFrame
	}

	return frame, nil
}

func writeHeader(buffer []byte, frameType FrameType, streamID uint32) {
	buffer[0] = byte(frameType)
	binary.BigEndian.PutUint32(buffer[1:], streamID)
}

File: packages/netbridge/src/protocol/frame.ts
// Wire format for the NetBridge multiplexer. Exactly one WebSocket message carries one frame, so the
// rendezvous in the control plane routes whole frames by stream without parsing the payload. A frame
// is a 1-byte type, a 4-byte stream id, and a type-specific tail:
//   Open  [1][streamId:4][port:2][host…]  — a new TCP stream to host:port (the name is resolved at the exit)
//   Data  [2][streamId:4][bytes…]         — payload for an open stream
//   Close [3][streamId:4]                 — the stream is finished in the sender's direction
export enum FrameType {
    Open = 1,
    Data = 2,
    Close = 3,
}

export type OpenFrame = {
    readonly type: FrameType.Open;
    readonly streamId: number;
    readonly host: string;
    readonly port: number;
};

export type DataFrame = {
    readonly type: FrameType.Data;
    readonly streamId: number;
    readonly payload: Buffer;
};

export type CloseFrame = {
    readonly type: FrameType.Close;
    readonly streamId: number;
};

export type Frame = OpenFrame | DataFrame | CloseFrame;

// Payload cap for a Data frame: the forwarder chunks a TCP stream into pieces no larger than this so a
// single WebSocket message never grows unbounded. The header (type + stream id) is not counted.
export const maxDataBytes = 64 * 1024;

const headerSize = 5;
const openPortSize = 2;
const maxStreamId = 0xffffffff;
const maxPort = 0xffff;

export class NetBridgeFrameError extends Error {}

export function encode(frame: Frame): Buffer {
    switch (frame.type) {
        case FrameType.Open:
            return encodeOpen(frame);
        case FrameType.Data:
            return encodeData(frame);
        case FrameType.Close:
            return encodeHeader(FrameType.Close, frame.streamId);
    }
}

export function decode(message: Buffer): Frame {
    if (message.length < headerSize) {
        throw new NetBridgeFrameError("frame is shorter than its header");
    }

    const type = message.readUInt8(0);
    const streamId = message.readUInt32BE(1);

    switch (type) {
        case FrameType.Open:
            return decodeOpen(message, streamId);
        case FrameType.Data:
            return { type: FrameType.Data, streamId, payload: message.subarray(headerSize) };
        case FrameType.Close:
            return { type: FrameType.Close, streamId };
        default:
            throw new NetBridgeFrameError(`unknown frame type ${type}`);
    }
}

function encodeOpen(frame: OpenFrame): Buffer {
    assertPort(frame.port);

    const host = Buffer.from(frame.host, "utf8");

    if (host.length === 0) {
        throw new NetBridgeFrameError("open frame requires a host");
    }

    const buffer = Buffer.allocUnsafe(headerSize + openPortSize + host.length);

    writeHeader(buffer, FrameType.Open, frame.streamId);
    buffer.writeUInt16BE(frame.port, headerSize);
    host.copy(buffer, headerSize + openPortSize);

    return buffer;
}

function encodeData(frame: DataFrame): Buffer {
    if (frame.payload.length > maxDataBytes) {
        throw new NetBridgeFrameError(`data frame exceeds ${maxDataBytes} bytes`);
    }

    const buffer = Buffer.allocUnsafe(headerSize + frame.payload.length);

    writeHeader(buffer, FrameType.Data, frame.streamId);
    frame.payload.copy(buffer, headerSize);

    return buffer;
}

function encodeHeader(type: FrameType, streamId: number): Buffer {
    const buffer = Buffer.allocUnsafe(headerSize);

    writeHeader(buffer, type, streamId);

    return buffer;
}

function writeHeader(buffer: Buffer, type: FrameType, streamId: number): void {
    assertStreamId(streamId);
    buffer.writeUInt8(type, 0);
    buffer.writeUInt32BE(streamId, 1);
}

function decodeOpen(message: Buffer, streamId: number): OpenFrame {
    if (message.length < headerSize + openPortSize) {
        throw new NetBridgeFrameError("open frame is missing its port");
    }

    const port = message.readUInt16BE(headerSize);
    const host = message.subarray(headerSize + openPortSize).toString("utf8");

    if (host.length === 0) {
        throw new NetBridgeFrameError("open frame is missing its host");
    }

    return { type: FrameType.Open, streamId, host, port };
}

function assertStreamId(streamId: number): void {
    if (!Number.isInteger(streamId) || streamId < 0 || streamId > maxStreamId) {
        throw new NetBridgeFrameError(`stream id out of range: ${streamId}`);
    }
}

function assertPort(port: number): void {
    if (!Number.isInteger(port) || port < 1 || port > maxPort) {
        throw new NetBridgeFrameError(`port out of range: ${port}`);
    }
}
